src/Compiler/Driver/ParseAndCheckInputs.fs

```diff
@@ -1338,11 +1351,44 @@ let CheckClosedInputSetFinish (declaredImpls: CheckedImplFile list, tcState) =
 
     tcState, declaredImpls, ccuContents
 
-let CheckClosedInputSet (ctok, checkForErrors, tcConfig, tcImports, tcGlobals, prefixPathOpt, tcState, inputs) =
+let CheckClosedInputSet (ctok, checkForErrors, tcConfig: TcConfig, tcImports, tcGlobals, prefixPathOpt, tcState, inputs) =
+    let isParallel = Environment.GetEnvironmentVariable("NOJAF_PARALLEL") |> String.IsNullOrWhiteSpace |> not
+    
     // tcEnvAtEndOfLastFile is the environment required by fsi.exe when incrementally adding definitions
     let results, tcState =
-        (tcState, inputs)
-        ||> List.mapFold (TypeCheckOneInputEntry(ctok, checkForErrors, tcConfig, tcImports, tcGlobals, prefixPathOpt))
+        if tcConfig.concurrentBuild && isParallel then
+            let results, tcState =
+                (tcState, inputs)
+                ||> List.mapFold (CheckOneInputEntrySkipImpl(ctok, checkForErrors, tcConfig, tcImports, tcGlobals, prefixPathOpt))
+
+            let inputs = Array.ofList inputs
+            let newResults = Array.ofList results
+            let results = Array.ofList results
+
+            (inputs, results)
+            ||> Array.zip
+            |> Array.mapi (fun i (input, (_, _, implOpt, _)) ->
+                match implOpt with
+                | None -> None
+                | Some impl ->
+                    match impl with
+                    | CheckedImplFile.CheckedImplFile (qualifiedNameOfFile = qualifiedNameOfFile
+                                                       contents = ModuleOrNamespaceContents.TMDefs []) ->
```

Comment:
Why the empty contents here?

Reply:
I guess it matches the results produced by this line https://github.com/dotnet/fsharp/pull/13737/files#diff-efdad7fa53b23cfadc0c316627cd7f8833d0a531db500cc7c220cc4fa0465ab2R1227 but it feels like it would be better to return a more explicit marker, in case some input with a signature file also happens to have empty defs